requirements/specifications/lifecycle/index.md

```diff
@@ -1,274 +1,365 @@
-a Firebolt compliant app, e.g. unregister for event listeners, clean up
-timers, post any final business analytics, etc.
+Apps in this state **SHOULD** use less than XXX megabytes of RAM.
 
-When an app is about to be unloaded the platform **MUST** transition it
-to this state. Apps **MUST** remain in this state until they notify the
-platform via Lifecycle.finished(), or are unloaded by the platform after
-appFinishedTimeout milliseconds have passed.
+**TODO**: do we want tiers of apps? per-app config?
 
-The platform **MUST NOT** ever put an app into any other state once it
-is in the unloading state unless it has been unloaded and then loaded
-again.
+When an app transitions to this state, the platform **MUST** dispatch
+the `Lifecycle.onSuspended` notification with the current state and
+previous states.
 
-### 2.6. Suspended
+### 2.5. Sleeping
 
 This state allows an app to be copied from memory to local storage and
 then terminated to save resources. Subsequently, the app may be copied
 back into memory, and resume in the same state.
 
-Apps **MUST** be able to subscribe to the Lifecycle.onSuspended
+Apps **MUST** be able to subscribe to the Lifecycle.onSleeping
 notification without errors regardless of whether the platform supports
 suspending apps. For platforms that do not support it the notification
 will just never fire.
 
-*If* a platform does not support suspending apps, then the following
+*If* a platform does not support the
+`xrn:firebolt:capability:lifecycle:sleep`, then the following
 requirements **MUST NOT** be fulfilled, even partially. A platform
 **MUST NOT** use the APIs documented here to implement an alternate,
-non-compliant version of the app suspending feature.
+non-compliant version of the app sleeping feature.
 
-*If* a platform supports suspending apps, then the following
-requirements **MUST** be fulfilled.
+*If* a platform supports the `xrn:firebolt:capability:lifecycle:sleep`
+capability, then the following requirements **MUST** be fulfilled.
 
-Apps **MUST** only enter this state from the inactive state.
+Apps **MUST** only enter this state from the `RUNNING` state, via
+the `suspend()` interface.
 
 *Immediately before* an app transitions to this state, the platform
-**MUST** dispatch the Lifecycle.onSuspended notification with the
+**MUST** dispatch the Lifecycle.onSleeping notification with the
 current state and previous states.
 
-The platform **MUST** save the app's memory space at this point, so that
-it may be unsuspended later. If storing the app\'s memory space fails
-for any reason, then the app **MUST** be treated as though it were
-unloaded.
+The platform **MUST** save the app's memory space at this point, so
+that it may be unsuspended later. If storing the app\'s memory space
+fails for any reason, then the app **MUST** be treated as though it
+were unloaded.
+
+Finally, the app and its container **MUST** be removed from memory
```

Comment:
This one may be more nuanced. I think some memory may remain even in sleep.

Reply:
Some sleep implementations also might do RAM compression instead of writing to disk.